### ISSUE TITLE
Display invisible character

### ISSUE THREAD

**Issue by noraj** [microsoft/vscode#58252]
## Description

Zero width characters can be used in some programs or to fingerprint people.

See [Zero-Width Characters - Invisibly fingerprinting text](https://www.zachaysan.com/writing/2017-12-30-zero-width-characters) and [Be careful what you copy: Invisibly inserting usernames into text with Zero-Width Characters](https://medium.com/@umpox/be-careful-what-you-copy-invisibly-inserting-usernames-into-text-with-zero-width-characters-18b4e6f17b66).

Those characters are not rendered in anyway, making hard to detect them.

So we need an option to:

- render invisible chars as their UTFx code (like in nvim <200b>
- or to render any invisible char as one arbitrary chosen char like for whitespace

[Invisible Character Visualizer](https://marketplace.visualstudio.com/items?itemName=ShaneRay.InvisibleCharacterVisualizer) seems to do that well so the feature can start from this plugin instead from scratch.
 
Second thing, the show whitespace feature of vscode is only displaying whitespaces and tabs but not CR and EOL like many editors.

For example in atom 

![](https://i.imgur.com/UFnP20p.png)

## Possible related issues and resources

- View invisible characters #31291
- Feature request: Show hidden characters like: space, tab, end of line, carriage return #18282
- Invisible Character Visualizer https://marketplace.visualstudio.com/items?itemName=ShaneRay.InvisibleCharacterVisualizer, but this is for Visual Studio not for VS Code

## Demo

VS Code (not even a clue there is something)

![](https://i.imgur.com/R9EMSWX.png)

neovim (nvim)

![](https://i.imgur.com/rJxbduR.png)

xxd

```
$ xxd test 
00000000: 436f 6e66 6964 656e 7469 616c 2041 6e6e  Confidential Ann
00000010: 6f75 6e63 656d 656e 743a 20e2 808c e280  ouncement: .....
00000020: 8be2 808b e280 8ce2 808b e280 8be2 808b  ................
00000030: e280 8ce2 808d e280 8ce2 808b e280 8be2  ................
00000040: 808c e280 8be2 808b e280 8be2 808b e280  ................
00000050: 8de2 808c e280 8be2 808b e280 8be2 808c  ................
00000060: e280 8ce2 808b e280 8ce2 808d e280 8ce2  ................
00000070: 808b e280 8be2 808c e280 8ce2 808c e280  ................
00000080: 8ce2 808b e280 8de2 808c e280 8be2 808b  ................
00000090: e280 8ce2 808b e280 8ce2 808b e280 8c54  ...............T
000000a0: 6869 7320 6973 2073 6f6d 6520 636f 6e66  his is some conf
000000b0: 6964 656e 7469 616c 2074 6578 7420 7468  idential text th
000000c0: 6174 2079 6f75 2072 6561 6c6c 7920 7368  at you really sh
000000d0: 6f75 6c64 6e27 7420 6265 2073 6861 7269  ouldn't be shari
000000e0: 6e67 2061 6e79 7768 6572 6520 656c 7365  ng anywhere else
000000f0: 2e0a                                     ..
```

**Comment by Instagit** [comment]
May I suggest having a setting to only show the whitespace and invisible characters in selections? Sublime Text has this option and it's really useful, because by default it doesn't get in the way, but you can easily see the characters when you want to.

I actually modified one of the Code color themes to do this [1], but making it a setting would be better.

[1] https://github.com/Instagit/vscode-theme-monokai-dark-soda-whitespace

![screenshot from 2018-11-08 14-34-23](https://user-images.githubusercontent.com/2681794/48201761-66c11f00-e363-11e8-82ac-a84e0d1eb1b1.png)

**Comment by noraj** [comment]
A setting to show invisible characters in selection but also a setting to show invisible characters globally.

Note that I'm talking about invisible characters as all non printable character not only whitespaces (space, tab, etc.).

**Comment by deadlyicon** [comment]
I just waisted 20m debugging a url problem because I had a Unicode Decimal Code &#8237; preceding my absolute url causing it to be treated as a relative path. 

Please allow us to enable rendering of these weird characters that no programmer ever wants in their code.

**Comment by Nicolab** [comment]
As in Atom would be great.
It's a must-have feature, I have trouble doing without.

**Comment by Phi-Li** [comment]
I've just experienced a compile error caused by Unicode character `U+202C` and `U+202D` hiding in pasted C code. It's really hard to locate and remove them. Since it's a common operation for programmers to paste code from somewhere else, I think there is a good reason to make displaying invisible character an internal feature, not just a third party plugin.

**Comment by espipj** [comment]
I have just experienced that zero width characters are not being "placed" at the right place in the editor. I am inserting a u200b invisible character and writing it to disk in PHP, when I open the written file the character seems to be in the wrong position (moving the cursor through all the characters in the file), however it shows correctly place when I open the same file with SublimeText 3

**Comment by hezedu** [comment]
https://github.com/eslint/eslint/issues/11752

**Comment by Nicolab** [comment]
I use Gremlins https://marketplace.visualstudio.com/items?itemName=nhoizey.gremlins
It works well

**Comment by nvmnghia** [comment]
Yes, do this please, I spent the whole afternoon to find out a f*cking `\u200b`

**Comment by noraj** [comment]
Plugins:

- https://marketplace.visualstudio.com/items?itemName=nachocab.highlight-dodgy-characters: very useful it highlight all all chars
- https://marketplace.visualstudio.com/items?itemName=wengerk.highlight-bad-chars: no handy as you must specify which char you want to be highlighted
- https://marketplace.visualstudio.com/items?itemName=nhoizey.gremlins: does highlight only a few chars such as whitespaces, etc. not all non printable

**Comment by cbrit** [comment]
@noraj Thank you for the "Highlight Dodgy Characters" suggestion. Solved my issue quickly and simply.

**Comment by saurabhpati** [comment]
> I just waisted 20m debugging a url problem because I had a Unicode Decimal Code ‭ preceding my absolute url causing it to be treated as a relative path.

Lucky guy, I wasted 2 days :-D

**Comment by fenying** [comment]
Any progress on this problem? I believe `0x200B` is a horrible character and should be highlight as a warning.
Now my solution is to write an ESLint plugin banning it.

**Comment by abdennour** [comment]
Waste for than 3 hours.

Kudos to this extension : https://marketplace.visualstudio.com/items?itemName=wengerk.highlight-bad-chars

**Comment by chrisinmtown** [comment]
👍 and +1 from me too.  I had some strange character in a file and VS code didn't show it, typing arrow keys mysteriously didn't move the cursor!  Should not have to resort to the good old 'od' command lol.

**Comment by shu-ming** [comment]
@noraj Thank you! Waste me a couple of hours to debug a rest call. The url looks good, but with the U+200c inside. With the extension you recommend, it is easy to figure out this kind of bugs.

**Comment by aakarshsingh** [comment]
+1

IntelliJ wins in terms of highlighting it. 

![image](https://user-images.githubusercontent.com/6056091/129924469-5f49c491-f315-400f-b7a1-c365facf32ac.png)

**Comment by sv-vmh-dev** [comment]
This is an important security feature and should be built into vscode with the ability to toggle on/off just like word wrap.
I would also add a shortcut to the status bar on the bottom that shows how many hidden characters are on the page.

See article: https://certitude.consulting/blog/en/invisible-backdoor/

**Comment by j824h** [comment]
Isn't this fixed with #137719?

**Comment by CatsMiaow** [comment]
https://code.visualstudio.com/updates/v1_63#_unicode-highlighting

**Comment by aakarshsingh** [comment]
Thank You 👍🏼 🥇

**Comment by hediet** [comment]
Fixed with #137508.

Feel free to open issues for use-cases described in this issue that are not covered by that PR.
